File: CSV2HTML_by_Java/csv2html/Writer.java
package csv2html;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.util.Calendar;
import java.util.List;
import utility.StringUtility;

/**
 * ライタ：情報のテーブルをHTMLページとして書き出す。
 */
public class Writer extends IO
{
	/**
	 * ライタのコンストラクタ。
	 * @param aTable テーブル
	 */
	public Writer(Table aTable)
	{
		super(aTable);

		return;
	}

	/**
	 * HTMLページを基にするテーブルからインデックスファイル(index.html)に書き出す。
	 */
	public void perform()
	{
		try
		{
			Attributes attributes = this.attributes();
			String fileStringOfHTML = attributes.baseDirectory() + attributes.indexHTML();
			File aFile = new File(fileStringOfHTML);
			FileOutputStream outputStream = new FileOutputStream(aFile);
			OutputStreamWriter outputWriter = new OutputStreamWriter(outputStream, StringUtility.encodingSymbol());
			BufferedWriter aWriter = new BufferedWriter(outputWriter);

			this.writeHeaderOn(aWriter);
			this.writeTableBodyOn(aWriter);
			this.writeFooterOn(aWriter);

			aWriter.close();
		}
		catch (UnsupportedEncodingException | FileNotFoundException anException) { anException.printStackTrace(); }
		catch (IOException anException) { anException.printStackTrace(); }

		return;
	}

	/**
	 * 属性リストを書き出す。
	 * @param aWriter ライタ
	 */
	public void writeAttributesOn(BufferedWriter aWriter)
	{
		try
		{
			Attributes attributes = this.attributes();

			// テーブル開始タグ
			aWriter.write("<table>");
			aWriter.newLine();

			// テーブルヘッダー行
			aWriter.write("<tr>");
			aWriter.newLine();

			// 各属性名をth要素として出力
			for (int index = 0; index < attributes.size(); index++)
			{
				String attributeName = attributes.at(index);
				aWriter.write("<th>" + IO.htmlCanonicalString(attributeName) + "</th>");
				aWriter.newLine();
			}

			aWriter.write("</tr>");
			aWriter.newLine();
		}
		catch (IOException anException) { anException.printStackTrace(); }

		return;
	}

	/**
	 * フッタを書き出す。
	 * @param aWriter ライタ
	 */
	public void writeFooterOn(BufferedWriter aWriter)
	{
		try
		{
			// テーブル終了タグ
			aWriter.write("</table>");
			aWriter.newLine();

			// BODY終了タグ
			aWriter.write("</body>");
			aWriter.newLine();

			// HTML終了タグ
			aWriter.write("</html>");
			aWriter.newLine();
		}
		catch (IOException anException) { anException.printStackTrace(); }

		return;
	}

	/**
	 * ヘッダを書き出す。
	 * @param aWriter ライタ
	 */
	public void writeHeaderOn(BufferedWriter aWriter)
	{
		try
		{
			Attributes attributes = this.attributes();

			// DOCTYPE宣言
			aWriter.write("<!DOCTYPE html>");
			aWriter.newLine();

			// HTML開始タグ
			aWriter.write("<html lang=\"ja\">");
			aWriter.newLine();

			// HEAD部分
			aWriter.write("<head>");
			aWriter.newLine();
			aWriter.write("<meta charset=\"" + StringUtility.encodingSymbol() + "\">");
			aWriter.newLine();
			aWriter.write("<title>" + IO.htmlCanonicalString(attributes.titleString()) + "</title>");
			aWriter.newLine();

			// スタイルを定義
			aWriter.write("<style>");
			aWriter.newLine();
			// Body: 白背景、serifフォント、余白設定
			aWriter.write("body { background-color: #ffffff; margin: 20px; padding: 10px; font-family: serif; font-size: 10pt; }");
			aWriter.newLine();
			// リンク: 下線あり、状態別の背景色
			aWriter.write("a { text-decoration: underline; color: #000000; }");
			aWriter.newLine();
			aWriter.write("a:link { background-color: #ffddbb; }");
			aWriter.newLine();
			aWriter.write("a:visited { background-color: #ccffcc; }");
			aWriter.newLine();
			aWriter.write("a:hover, a:active { background-color: #dddddd; }");
			aWriter.newLine();
			// 画像: 枠なし、中央揃え
			aWriter.write("img { border: 0px; vertical-align: middle; }");
			aWriter.newLine();
			// テーブル: 白枠線、幅100%
			aWriter.write("table { border-collapse: collapse; border: 1px solid #ffffff; background-color: #ffffff; width: 100%; }");
			aWriter.newLine();
			// ヘッダセル: ピンク背景、中央揃え、白枠線
			aWriter.write("th { background-color: #ffddee; text-align: center; padding: 4px; border: 1px solid #ffffff; }");
			aWriter.newLine();
			// データセル: 基本スタイル、白枠線
			aWriter.write("td { padding: 4px; border: 1px solid #ffffff; text-align: center; }");
			aWriter.newLine();
			// 行ごとの色分け（奇数行: 青、偶数行: 黄色）
			aWriter.write("tr:nth-child(odd) td { background-color: #ddeeff; }");  // 奇数行: 青
			aWriter.newLine();
			aWriter.write("tr:nth-child(even) td { background-color: #ffffcc; }"); // 偶数行: 黄色
			aWriter.newLine();
			// 見出し（h1）のスタイル
			aWriter.write("h1 { font-size: 16pt; margin-bottom: 10px; background-color: #EBEBEB; padding: 4px; }");
			aWriter.newLine();
			aWriter.write("</style>");
			aWriter.newLine();
			aWriter.write("</head>");
			aWriter.newLine();

			// BODY開始タグ
			aWriter.write("<body>");
			aWriter.newLine();
			aWriter.write("<h1>" + IO.htmlCanonicalString(attributes.captionString()) + "</h1>");
			aWriter.newLine();
		}
		catch (IOException anException) { anException.printStackTrace(); }

		return;
	}

	/**
	 * ボディを書き出す。
	 * @param aWriter ライタ
	 */
	public void writeTableBodyOn(BufferedWriter aWriter)
	{
		this.writeAttributesOn(aWriter);
		this.writeTuplesOn(aWriter);

		return;
	}

	/**
	 * タプル群を書き出す。
	 * @param aWriter ライタ
	 */
	public void writeTuplesOn(BufferedWriter aWriter)
	{
		try
		{
			Attributes attributes = this.attributes();
			int imageIndex = attributes.indexOfImage();

			// 各タプル（データ行）を出力
			for (Tuple aTuple : this.tuples())
			{
				aWriter.write("<tr>");
				aWriter.newLine();

				// 各値をtd要素として出力
				List<String> values = aTuple.values();
				for (int index = 0; index < values.size(); index++)
				{
					String value = values.get(index);
					// 画像列はHTMLタグを含むのでエスケープしない
					String cellContent = (index == imageIndex) ? value : IO.htmlCanonicalString(value);
					aWriter.write("<td>" + cellContent + "</td>");
					aWriter.newLine();
				}

				aWriter.write("</tr>");
				aWriter.newLine();
			}
		}
		catch (IOException anException) { anException.printStackTrace(); }

		return;
	}
}

File: CSV2HTML_by_Python/csv2html/writer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""CSV2HTML：総理大臣と徳川幕府の情報「CSV」からWebページ「HTML」を生成。"""
__author__ = 'AOKI Atsushi'
__version__ = '1.0.7'
__date__ = '2021/01/10 (Created: 2016/01/01)'

import os

from csv2html.io import IO

# pylint: disable=R0201
# R0201: Method could be a function (no-self-use)

class Writer(IO):
	"""ライタ：情報のテーブルをHTMLページとして書き出す。"""

	def __init__(self, output_table):
		"""ライタのコンストラクタ。HTMLページを基にするテーブルを受け取る。"""

		super().__init__(output_table)
		(lambda x: x)(output_table) # NOP

	def perform(self):
		"""HTMLページを基にするテーブルから、インデックスファイル(index_html)に書き出す。"""

		class_attributes = self.attributes().__class__
		base_directory = class_attributes.base_directory()
		index_html = class_attributes.index_html()

		html_filename = os.path.join(base_directory, index_html)
		with open(html_filename, 'w', encoding='utf-8') as a_file:
			self.write_header(a_file)
			self.write_body(a_file)
			self.write_footer(a_file)

	def write_body(self, file):
		"""ボディを書き出す。つまり、属性リストを書き出し、タプル群を書き出す。"""

		file.write("<body>\n")
		file.write(f"<h1>{self.attributes().caption_string()}</h1>\n")
		
		file.write("<table>\n")
		
		# 属性リスト（ヘッダー）を書き出す
		file.write("<tr>\n")
		for name in self.attributes().names():
			file.write(f"<th>{self.html_canonical_string(name)}</th>\n")
		file.write("</tr>\n")
		
		# タプル群（データ）を書き出す
		tuples = self.table().tuples()
		
		# 画像列のインデックスを取得
		image_index = -1
		try:
			image_index = self.attributes().keys().index("image")
		except ValueError:
			pass

		for i, a_tuple in enumerate(tuples):
			file.write("<tr>\n")
			values = a_tuple.values()
			for j, value in enumerate(values):
				# 画像列はHTMLタグを含むのでエスケープしない
				if j == image_index:
					file.write(f"<td>{value}</td>\n")
				else:
					file.write(f"<td>{self.html_canonical_string(value)}</td>\n")
			file.write("</tr>\n")
			
		file.write("</table>\n")
		file.write("</body>\n")

	def write_footer(self, file):
		"""フッタを書き出す。"""

		file.write("</html>\n")

	def write_header(self, file):
		"""ヘッダを書き出す。"""

		file.write("<!DOCTYPE html>\n")
		file.write('<html lang="ja">\n')
		file.write("<head>\n")
		file.write('<meta charset="utf-8">\n')
		file.write(f"<title>{self.attributes().title_string()}</title>\n")
		
		file.write("<style>\n")
		file.write("body { background-color: #ffffff; margin: 20px; padding: 10px; font-family: serif; font-size: 10pt; }\n")
		file.write("a { text-decoration: underline; color: #000000; }\n")
		file.write("a:link { background-color: #ffddbb; }\n")
		file.write("a:visited { background-color: #ccffcc; }\n")
		file.write("a:hover, a:active { background-color: #dddddd; }\n")
		file.write("img { border: 0px; vertical-align: middle; }\n")
		file.write("table { border-collapse: collapse; border: 1px solid #ffffff; background-color: #ffffff; width: 100%; }\n")
		file.write("th { background-color: #ffddee; text-align: center; padding: 4px; border: 1px solid #ffffff; }\n")
		file.write("td { padding: 4px; border: 1px solid #ffffff; text-align: center; }\n")
		file.write("tr:nth-child(odd) td { background-color: #ddeeff; }\n")
		file.write("tr:nth-child(even) td { background-color: #ffffcc; }\n")
		file.write("h1 { font-size: 16pt; margin-bottom: 10px; background-color: #EBEBEB; padding: 4px; }\n")
		file.write("</style>\n")
		
		file.write("</head>\n")
